### PR TITLE
feat: add multi message

### PR DESCRIPTION
**Pull request by ibrizsabin**
### Commit Message

```bash
type: commit message
```

see the [guidelines](https://github.com/icon-project/community/blob/main/guidelines/technical/software-development-guidelines.md#commit-messages) for commit messages.

### Changelog Entry

```bash
version: <log entry>
```

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have documented my code in accordance with the [documentation guidelines](https://github.com/icon-project/community/blob/main/guidelines/technical/software-development-guidelines.md#documentation)
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have run the unit tests
- [ ] I only have one commit (if not, squash them into one commit).
- [ ] I have a descriptive commit message that adheres to the [commit message guidelines](https://github.com/icon-project/community/blob/main/guidelines/technical/software-development-guidelines.md#commit-messages)

> Please review the [CONTRIBUTING.md](/CONTRIBUTING.md) file for detailed contributing guidelines.

### CODE REVIEW

File: contracts/cosmwasm-vm/cw-xcall/src/send_call_message.rs [hunk 1]
```diff
@@ -17,47 +22,84 @@ impl<'a> CwCallService<'a> {
...
-                CallRequest::new(caller.clone(), to.clone(), sources, rollback_data, false);
+        if envelope.message.should_persist() {
+            let rollback_data = envelope.message.rollback().unwrap();
+            let request = CallRequest::new(
+                caller.clone(),
+                to.clone(),
+                envelope.sources,
+                rollback_data,
+                false,
+            );
```
Comment (AntonAndell) [inline contracts/cosmwasm-vm/cw-xcall/src/send_call_message.rs:88]: Not necessarily valid depending on future messages. 
Should persist has nothing to do with if it should have a response?

File: contracts/cosmwasm-vm/cw-xcall/src/send_call_message.rs [hunk 2]
```diff
@@ -17,47 +22,84 @@ impl<'a> CwCallService<'a> {
...
             sequence_no,
-            need_response,
-            data.to_vec(),
-            destinations,
+            envelope.message.msg_type().clone(),
+            envelope.message.data(),
+            envelope.destinations,
         );
 
+        let need_response = call_request.need_response();
```
Comment (AntonAndell) [inline contracts/cosmwasm-vm/cw-xcall/src/send_call_message.rs:101]: need_response() this would be fine but this method might need more additions as we add more messages with responses. Would this still be more clear. Since a new message would require you to edit the callrequest file

File: contracts/cosmwasm-vm/cw-xcall/src/execute_call.rs [hunk 3]
```diff
@@ -82,7 +82,7 @@ impl<'a> CwCallService<'a> {
             cosmwasm_std::SubMsgResult::Ok(_res) => {
                 let code = CallServiceResponseType::CallServiceResponseSuccess.into();
```
Comment (AntonAndell) [inline contracts/cosmwasm-vm/cw-xcall/src/execute_call.rs:84]: Where is the message specific execution?

**Review by AntonAndell — CHANGES_REQUESTED** [review]
Looks like it would work with the same flow internally. But implementing a new message type seems complicated and the flow is somewhat hard to follow. You can't really see what each message does clearly

File: contracts/cosmwasm-vm/cw-xcall/src/execute_call.rs (hunk 3 above)
Comment (ibrizsabin) [reply contracts/cosmwasm-vm/cw-xcall/src/execute_call.rs:84]: on execution only case currently is whether  request needs response or not which is handled. other cases can also be added similarly without needing to know the message type explicitly.

File: contracts/cosmwasm-vm/cw-xcall/src/send_call_message.rs (hunk 2 above)
Comment (ibrizsabin) [reply contracts/cosmwasm-vm/cw-xcall/src/send_call_message.rs:101]: ya thats a valid concern since we dont want to modify storage object but the flag need_response is just copy from the actual message type so unless we have aditional flags in message flow itself that needs to be handled in execution we wont have any problem and when we have such flags we would need to modify storage anyway.

File: contracts/cosmwasm-vm/cw-xcall/src/send_call_message.rs (hunk 1 above)
Comment (ibrizsabin) [reply contracts/cosmwasm-vm/cw-xcall/src/send_call_message.rs:88]: ya this is a naming mistake. it should rather be has_rollback  or should_log . I had understood persist differently. let me fix this.

File: contracts/cosmwasm-vm/cw-xcall-lib/src/message/envelope.rs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_encoding_decoding_envelope_call_message() {
+        // Create a sample Envelope
+        let message = AnyMessage::CallMessage(CallMessage {
+            msg_type: MessageType::BasicMessage,
```
[the remark this answers is not in the document]
Comment (AntonAndell) [reply contracts/cosmwasm-vm/cw-xcall-lib/src/message/envelope.rs:81]: It seems you just used what was in java at the time, but lets follow what is in the document PR

**Review by AntonAndell — APPROVED** [review]

File: contracts/cosmwasm-vm/cw-xcall/tests/test_handle_call_message.rs [hunk 5]
```diff
@@ -9,7 +9,7 @@ use cw_xcall::{
     state::{CwCallService, EXECUTE_CALL_ID},
     types::{call_request::CallRequest, request::CSMessageRequest},
 };
-use cw_xcall_lib::network_address::NetworkAddress;
+use cw_xcall_lib::{message::msg_type::MessageType, network_address::NetworkAddress};
```
[the remark this answers is not in the document]
Comment (ibrizsabin) [reply contracts/cosmwasm-vm/cw-xcall/tests/test_handle_call_message.rs:12]: ya let me align it with the doc

File: contracts/cosmwasm-vm/cw-xcall-lib/src/message/call_message.rs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    pub data: Vec<u8>,
+}
+
+impl Encodable for CallMessage {
+    fn rlp_append(&self, stream: &mut RlpStream) {
+        stream
+            .begin_list(2)
+            .append(&Into::<u8>::into(self.msg_type.clone()))
+            .append(&self.data);
+    }
```
[the remark this answers is not in the document]
Comment (ibrizsabin) [reply contracts/cosmwasm-vm/cw-xcall-lib/src/message/call_message.rs:17]: ya i think its redundant to have msg type here as well , (msg_typ+msg_bytes) is represented as AnyMessage which should be enough. I think i tried to emulate the BaseMessage parent class but is not necessary.

File: contracts/cosmwasm-vm/cw-xcall/src/send_call_message.rs [hunk 7]
```diff
@@ -17,47 +22,84 @@ impl<'a> CwCallService<'a> {
...
 
         if confirmed_sources.is_empty() {
             let default = self.get_default_connection(deps.as_ref().storage, to.nid())?;
             confirmed_sources = vec![default.to_string()]
         }
 
-        if need_response {
-            let request =
-                CallRequest::new(caller.clone(), to.clone(), sources, rollback_data, false);
+        if envelope.message.rollback().is_some() {
```
[the remark this answers is not in the document]
Comment (ibrizsabin) [reply contracts/cosmwasm-vm/cw-xcall/src/send_call_message.rs:76]: we can do it both ways ... in flow we can stick with the flags the message types are to simplify the flow to users so that they dont need to worry explicity about the flags.
